### ISSUE TITLE
Redefinition of structs in .h and .cc requires a whole extra header file

### ISSUE THREAD

**Issue by adetaylor** [dtolnay/cxx#237]
In the following example I've tried to move a pre-existing C++ struct such that it can be used in Rust code too.

```rust
#[cxx::bridge(namespace = base)]
pub mod ffi {
    struct DeviceUserDomainJoinState { // previously defined in C++
        device_joined: bool,
        user_joined: bool,
    }   

    extern "C" {
        include!("base/enterprise_util.h"); // it was previously defined in here

        fn AreDeviceAndUserJoinedToDomain() -> DeviceUserDomainJoinState;
    }
}   

fn get_user_state() -> bool {
    let state = ffi::AreDeviceAndUserJoinedToDomain();
    state.user_joined
}   
```

Existing .h file changes:

```diff
diff --git a/base/enterprise_util.h b/base/enterprise_util.h
index e14116e2694b..a61e6c74fab2 100644
--- a/base/enterprise_util.h
+++ b/base/enterprise_util.h
@@ -7,6 +7,7 @@
 
 #include "base/base_export.h"
 #include "build/build_config.h"
+#include "base/rs/enterprise_util.h" // include cxxbridge-generated file
 
 namespace base {
 
@@ -49,10 +50,6 @@ enum class MacDeviceManagementStateNew {
 BASE_EXPORT MacDeviceManagementStateNew IsDeviceRegisteredWithManagementNew();
 
 // Returns whether the device and/or the current user is enrolled to a domain.
-struct DeviceUserDomainJoinState {
-  bool device_joined;
-  bool user_joined;
-};
 BASE_EXPORT DeviceUserDomainJoinState AreDeviceAndUserJoinedToDomain();
 
 #endif  // OS_MACOSX
```

It doesn't build:

```
gen/base/rs/enterprise_util.cc:7:8: error: redefinition of 'DeviceUserDomainJoinState'
struct DeviceUserDomainJoinState final {
```

This is because the `struct` is defined in both the generated .h file and the generated .cc file, so I have this problem:

* Generated `gen/base/rs/enterprise_util.cc` defines `DeviceUserDomainJoinState` and includes:
  * Hand-written `base/enterprise_util.h` which declares the `AreDeviceAndUserJoinedToDomain` function we need, and includes:
    * Generated `gen/base/rs/enterprise_util.h`, which also defines `DeviceUserDomainJoinState`.

As far as I can see, the only way to break this cycle with the current `cxxbridge` would be to:
* Create a new header, say, `enterprise_util_rust_interop.h`, containing:

```cpp
#pragma once
#include "base/base_export.h"

namespace base {
struct DeviceUserDomainJoinState;
BASE_EXPORT DeviceUserDomainJoinState AreDeviceAndUserJoinedToDomain();
}
```
* Modify the `.rs` file to `include!("base/enterprise_util_rust_interop.h");`

It would be great to be able to remove this extra interop header. Could we make this less awkward by:
1. Only ever defining structs once, in the generated .h file
2. Including the generated .h file from the generated .cc file?

There must be a reason you didn't do things this way in the first place, so let me know what I'm missing!

**Comment by dtolnay** [comment]
Yep, this needs to be fixed. Someone else had just run into this last week and I had filed #230 to follow up.

> There must be a reason you didn't do things this way in the first place, so let me know what I'm missing!

Currently it's somewhat convenient that the code generator doesn't need to care where you place the generated code or how include paths work. If we had to make the generated .cc file include the generated .h file the caller would need to tell us what include path to use there, either by writing it in the source Rust file which is somewhat strange, or as a command line argument which is okay but I was hoping to get away with avoiding (since the code generator has all the information to manually inline the include instead).

The alternative fix indicated in #230 is to emit an ifdef around the struct definition in both generated files. Do you see any downsides of that for your use case?

**Comment by adetaylor** [comment]
It looks like my search-fu needs to be improved :) Yes that would be fine. I'll close this since I think it's fine for it to be a duplicate.
